### PR TITLE
Make embedded documents be defined as real <embedded-document>

### PR DESCRIPTION
**Pull request by hairmare**
Basically this contains work on making all doctrine `<embed-one>` and `<embed-many>` relationships point to a document that has been defined as `<embedded-document>`. This is #298 and [EVO-1099](https://issue.swisscom.ch/browse/EVO-1099).

When done this should fix #298. It has been quite tricky to find a solution that does not break BC everywhere, but I believe that this is almost there.

To check this you can do the following while ignoring build failures.

```bash
rm -rf src/GravionDyn/ && sh dev-cleanstart.sh
rm -rf app/cache/ && vendor/bin/phpunit src/Graviton/CoreBundle/Tests/Controller/EmbeddingDocumentsTest.php
```

This PR contains my unrebased work on the theme and will need rebasing after everything works. I was planning on cleaning this up and squashing it before making a pr. I'm making one now since I'll be traveling and this can't wait until I return.

* [x] make generator generate both embedded and regular documents (using the suffix `Embedded` for embeddable variants)
* [x] make docs generated from json-defs use embedded versions of documents throughout
* [x] make sure that the `id` field in embedded documents is a simple string and not a id required by docgtrine
* [x] add a replacement id (`realId`) so doctrine still works
* [x] hide the `realId` using the serializer so it never gets exposed
* [x] change direction of class relationships from `DocumentEmbedded extends Document` to `Document extends DocumentEmbedded` since doctrine/mongodb-odm > 1.0.0 does not support disabling id fields in subclasses (note that composer.json is still locked to a beta, which it should not be in the end)
* [ ] change generator so `id` field only gets added to `Document` and not `DocumentEmbedded`
* [ ] consider getting rid of `realId` in `Document` by moving it to `DocumentEmbedded` (possibly by way of a mapped superclass containing everything else)
* [x] make fixtures loading also work with embedded structures (they always write references as it seems)
* [x] make sure swagger generation works again (This broke and I haven't found the time to look into it)
* [ ] make it work for `deep.nested.stuff` (probably has to do with rewriting `class:GravitonDyn\...[]` style types to `class:GravitonDyn\...Embedded[]`, I've concentrated on simple cases so far)
* [ ] check if `PersonContactEmbedded` and `HashEmbedded` are really needed (I had lots of stuff on my working copy when adding them and they might be obsolete again, `PersonContactEmbedded` will most likely not work due to how it's id is defined (the doctrine/mongodb-odm > 1.0.0 issue))
* [x] check that integrations tests work again
* [x] fix generator tests (they are mostly unit-tests that check how many times we try to write a file with twig)

### CODE REVIEW

**Comment by narcoticfresh** [comment]
@mrix we're coming closer to a solution of this problem.. your `EmbeddingDocumentsTest` is all green now..
can u tell me if `EmbeddingDocumentsTest` here reflects all checks that are necessary or is something missing?

it's important to note that in my opinion, that Documents that are no longer referenced should *not* be deleted..

once I squash all remaining issues and tests are green again (and if your tests are complete), I'll move your test definitions to the `TestBundle` and we're going to keep `EmbeddingDocumentsTest`.

**Comment by mrix** [comment]
@narcoticfresh Yes. These tests cover this bug completely. But I have a suggestion.

It would be great if we would have real inheritance of Doctrine mapping, Serializer rules and Validation constraints. And I think generating of embedded documents on demand is preferred.

For example:

```
/////////////////////////////////////////////////////////////////
// we have the following models
/////////////////////////////////////////////////////////////////
// embedded
agenda = {
    name: string
}
checklist = {
    id: string
    name: string
}

// documents
code = {
    id: string,
    name: string
}
customer = {
    id: string,
    title: code
}
consultation = {
    id: string,
    status: code,
    customer: customer,
    agenda: agenda,
    checklist: checklist
}



/////////////////////////////////////////////////////////////////
// base classes for each document
// base classes have `id` field but it's not mapped
// + doctrine @superclass mapping
// + generate validation rules
// + serializer rules
//
//
// extend *__base classes
// + doctrine @document mapping
// + generate validation rules for extended classes
// + serializer rules for extended documents
/////////////////////////////////////////////////////////////////
agenda__base = {
    name: string
}
checklist__base = {
    id: string,
    name: string
}

code__base = {
    name: string
}
code: code__base = {
    id: string
}

// we copy "code" to "customer" namespace on demand
customer_code: code__base = {
    _id: hidden,
    id: string
}
customer__base = {
    title: customer_code
}
customer: customer__base = {
    id: string
}

// we copy "code", "customer", "agenda" and "checklist" to "consultation" namespace
consultation_agenda: agenda__base = {
    _id: hidden
}
consultation_checklist: checklist__base = {
    _id: hidden
}
consultation_code: code__base = {
    _id: hidden,
    id: string
}
consultation_customer: customer__base = {
    _id: hidden,
    id: string
}
consultation__base = {
    status: consultation_code,
    customer: consultation_customer,
    agenda: consultation_agenda,
    checklist: consultation_checklist
}
consultation: consultation__base = {
    id: string
}
```

The described above schema requires `parentClass` field in JSON definition.
`JsonDefinitionRel::getJsonDefinition()` should be added like in `JsonDefinitionHash`. The new method will generate simple JSON definition with `parentClass` field.

Of course we can use `parentClass` manually:

```
extended_code: code = {
    extra: string
}
```

**Comment by mrix** [comment]
@narcoticfresh Today or tomorrow I will deploy this branch to our instance and check it with real client.

**Comment by mrix** [comment]
@narcoticfresh I got the error when I tried to deploy this branch:
```
[Symfony\Component\Debug\Exception\ContextErrorException]                                                                                                                                             
Catchable Fatal Error: Argument 1 passed to GravitonDyn\FinancingPositionBundle\Document\FinancingPositionBase::setCurrency() must be an instance of GravitonDyn\CodeBundle\Document\CodeEmbedded, i  
nstance of MongoDBODMProxies\__CG__\GravitonDyn\CodeBundle\Document\Code given, called in .../src/GravitonDyn/FinancingPositionBundle/DataFixtures/MongoDB/Lo  
adFinancingPositionData.php on line 74 and defined    
```

**Comment by narcoticfresh** [comment]
@mrix you need to remove  src/GravitonDyn before manually u build.. 

Regarding your ideas.. We have no time for big rewrites or whatever.. we just need to get it working in a timely fashion - we can always do this better in the future..

Currently, I still need to complete the validation.xml part, but i had no more time to work on this

**Comment by narcoticfresh** [comment]
@mrix btw, I have a mapped supperclass now and I tried to make serializer config per-instance with splitted up fields.. we have inheritance there now.. but the serializer didn't properly apply those things..

**Comment by mrix** [comment]
> @mrix you need to remove src/GravitonDyn before manually u build..

It still does not work.

I only replaced Graviton version with `"graviton/graviton": "dev-feature/298-make-embedded-documents-be-real-embedded-copies"` in API wrapper:
```
$ cat composer.lock | grep -B1 -A2 '"graviton/graviton"'
        {
            "name": "graviton/graviton",
            "version": "dev-feature/298-make-embedded-documents-be-real-embedded-copies",
            "source": {
--
    "stability-flags": {
        "graviton/graviton": 20,
        "libgraviton/swagger-ui": 20
    },
```

Prepare:
```
$ export SYMFONY_ENV=dev
$ export COMPOSER_CMD=composer
$
$ composer install
\Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Nothing to install or update
Generating autoload files
> Incenteev\ParameterHandler\ScriptHandler::buildParameters
Updating the "vendor/graviton/graviton/app/config/parameters.yml" file
Updating the "app/config/deploy.yml" file
> Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::buildBootstrap
> Graviton\CoreBundle\Composer\ScriptHandler::generateVersionYml
> Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::clearCache
Clearing the cache for the dev environment with debug true
> Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::installRequirementsFile
> Graviton\SwaggerBundle\Composer\ScriptHandler::copySwagger
> Graviton\SwaggerBundle\Composer\ScriptHandler::generateSwaggerJson
$
$ rm -Rf vendor/graviton/graviton/app/cache/*
$ rm -Rf vendor/graviton/graviton/app/logs/*
$ rm -Rf vendor/graviton/app/bootstrap.php.cache
$ rm -Rf vendor/graviton/graviton/src/GravitonDyn/
$
$ vendor/graviton/graviton/app/console -n graviton:generate:dynamicbundles --json > /dev/null
$
$ rm -Rf vendor/graviton/app/bootstrap.php.cache
$ composer run-script post-update-cmd
> Incenteev\ParameterHandler\ScriptHandler::buildParameters
Updating the "vendor/graviton/graviton/app/config/parameters.yml" file
Updating the "app/config/deploy.yml" file
> Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::buildBootstrap
> Graviton\CoreBundle\Composer\ScriptHandler::generateVersionYml
> Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::clearCache
Clearing the cache for the dev environment with debug true
> Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::installRequirementsFile
> Graviton\SwaggerBundle\Composer\ScriptHandler::copySwagger
> Graviton\SwaggerBundle\Composer\ScriptHandler::generateSwaggerJson
```

Load fixtures:
```
$ vendor/graviton/graviton/app/console doctrine:mongodb:fixtures:load
...
[Symfony\Component\Debug\Exception\ContextErrorException]
Catchable Fatal Error: Argument 1 passed to GravitonDyn\FinancingPositionBundle\Document\FinancingPositionBase::setCurrency() must be an instance of GravitonDyn\CodeBundle\Document\CodeEmbedded, i  
nstance of MongoDBODMProxies\__CG__\GravitonDyn\CodeBundle\Document\Code given, called in /sync/graviton-api-szkb-bap/vendor/graviton/graviton/src/GravitonDyn/FinancingPositionBundle/DataFixtures/  
MongoDB/LoadFinancingPositionData.php on line 74 and defined
...
```

File: src/Graviton/GeneratorBundle/Resources/skeleton/document/Document.mongodb.xml.twig [hunk 1]
```diff
@@ -4,16 +4,28 @@
       xmlns:gedmo="http://gediminasm.org/schemas/orm/doctrine-extensions-mapping"
       xsi:schemaLocation="http://doctrine-project.org/schemas/odm/doctrine-mongo-mapping http://doctrine-project.org/schemas/odm/doctrine-mongo-mapping.xsd">
 
-    {% if json is defined and json.isSubDocument() == true %}
-        {% set docType = "embedded-document" %}
+    <mapped-superclass name="{{ base }}Document\{{ document }}Base">
```
Comment (mrix) [inline src/Graviton/GeneratorBundle/Resources/skeleton/document/Document.mongodb.xml.twig]: > @mrix btw, I have a mapped supperclass now

So why is `<mapped-superclass>` empty? You can put all fields except `id` here. Then `<embedded-document>` and `<document>` will contain only `id` and `realId` fields.
Symfony validator also supports inheritance. But I'm not sure about JMS serializer.

**Comment by narcoticfresh** [comment]
@mrix yes, i thought so.. it also seems to work for `id`.. but then I had the problem of `$ref` rendered as `ref` (fieldName).. so then it seemed that `exposeAs` somehow didn't work if it was on the baseclass - the Serializer didn't handle that correctly.. and as time is running out, I'm leaving this as-is.. we always can arrange some time to optimize that..

to your problems: the fixture load has problems if a service definition contains fixtures that define a relation as `embed` but then try to load a `$ref: .., $id..` structure into that field.. i need to go through all those fixtures.. I'm doing that currently..

**Comment by narcoticfresh** [comment]
replaced by #321 to have a cleaner git history and less comment noise..

**Pull request closed without merging**